### PR TITLE
Fix two vanilla bugs that kill player via the pinched code

### PR DESCRIPTION
**Pull request by ds-sloth**
Developer-facing changes:
- Multiple Pinched values (5 integers) replaced with a single bitfield (`PinchedInfo_t`), with full mememu support. Saves ~100kb of RAM and makes NPCs more cache-friendly. Small CPU expense, but the pinched info isn't accessed in hot loops.

User-facing changes:
- Fixed vanilla bug where player could be prematurely crushed by a sloped ceiling
 

- Fixed vanilla bug where player could be crushed by a horizontally (not vertically) moving ceiling (these examples are from the same gameplay recording script, in which the Mode 3 player dies and the Mode 1 player survives)
 

I'd like to include this in 1.3.6.1, but, we should probably make a gameplay feature freeze soon because the main branch now includes the 1.3.7 performance updates.

[pinch-tests.zip](https://github.com/Wohlstand/TheXTech/files/10574388/pinch-tests.zip)

### CODE REVIEW

File: src/script/luna/mememu.cpp [hunk 1]
```diff
@@ -1499,7 +1562,25 @@ class NPCMemory final : public NpcParent
...
+            case 0x0C: // Pinched2
+                return valueToMem((int)obj->Pinched.Left2, ftype);
+            case 0x0E: // Pinched3
+                return valueToMem((int)obj->Pinched.Top3, ftype);
+            case 0x10: // Pinched4
+                return valueToMem((int)obj->Pinched.Right4, ftype);
+            case 0x12: // MovingPinched
+                return valueToMem((int)obj->Pinched.Moving, ftype);
+            default:
+                // use default handler
```
Comment (Wohlstand) [inline src/script/luna/mememu.cpp:1580]: I guess, the warning message should be added here in any cases of attempts to hack bytes in this area.

File: src/script/luna/mememu.cpp [hunk 2]
```diff
@@ -1524,6 +1605,40 @@ class NPCMemory final : public NpcParent
...
+                obj->Pinched.Top3 = in;
+                return;
+            case 0x10: // Pinched4
+                obj->Pinched.Right4 = in;
+                return;
+            case 0x12: // MovingPinched
+                obj->Pinched.Moving = in;
+                return;
+            default:
+                // use default handler
```
Comment (Wohlstand) [inline src/script/luna/mememu.cpp:1638]: Same as above:
I guess, the warning message should be added here in any cases of attempts to hack bytes in this area.

File: src/script/luna/mememu.cpp (hunk 1 above)
Comment (ds-sloth) [reply src/script/luna/mememu.cpp:1580]: Good idea, I'll fix that now in both cases

**Comment by 0lhi** [comment]
This seems wrong:

Scr_2023-02-03_01-14-37

**Comment by ds-sloth** [comment]
Strangely enough, I think that's actually the Mode 3 behavior too. I'd rather just fix the unexpected deaths, not every bug related to the slope code (which is certainly not ideal).

**Comment by 0lhi** [comment]
> Strangely enough, I think that's actually the Mode 3 behavior too. I'd rather just fix the unexpected deaths, not every bug related to the slope code (which is certainly not ideal).

True, I could reproduce in Mode 3 with God Mode :o 

Scr_2023-02-10_04-10-25

I'm not sure how I feel about this. By fixing one issue, another gets more prominence. It might be better to leave it unchanged.

**Comment by ds-sloth** [comment]
It seems like the undesirable behavior is exclusive to the normal blocks moving downwards while the player is on a sloped floor.

When the sloped floor moves upwards, both vanilla and TheXTech allow the player to clip through the ceiling.



I would be happy to revert the changes for the sloped floor case. I imagine it's a lot less common than the sloped ceiling case, and it's also glitchier in the vanilla game already.

**Comment by ds-sloth** [comment]
I reverted the sloped floor case, and also retargeted this change to 1.3.7.

We might want to revisit the behavior of this compat flag before releasing that version if somebody notices another unusual case during testing.

**Comment by ds-sloth** [comment]
@Wohlstand @0lhi does this seem good to merge? We can revise the behavior more later if anything unexpected emerges before the 1.3.7 release, but I think the essential issues have been discussed now.

**Comment by 0lhi** [comment]
Will this be excluded from the 1.3.6.1 Release?

Merge should be okay if you make an issue to do more testing before 1.3.7.

**Comment by ds-sloth** [comment]
I would like to exclude this from 1.3.6.1, yes.

**Comment by Wohlstand** [comment]
Seems to be fine to merge, but still need a further testing to ensure nothing significant got broken.

**Pull request merged**